icu4c/source/test/fuzzer/calendar_fuzzer.cpp

```diff
@@ -0,0 +1,125 @@
+// © 2023 and later: Unicode, Inc. and others.
+// License & terms of use: http://www.unicode.org/copyright.html
+
+// Fuzzer for ICU Calendar.
+
+#include <cstring>
+
+#include "fuzzer_utils.h"
+
+#include "unicode/calendar.h"
+#include "unicode/localebuilder.h"
+#include "unicode/locid.h"
+
+icu::TimeZone* CreateRandomTimeZone(uint16_t rnd) {
+    icu::Locale und("und");
+    UErrorCode status = U_ZERO_ERROR;
+    std::unique_ptr<icu::StringEnumeration> enumeration(
+        icu::TimeZone::createEnumeration(status));
+    if (U_SUCCESS(status)) {
+        int32_t count = enumeration->count(status);
+        if (U_SUCCESS(status)) {
+            int32_t i = rnd % count;
+            const icu::UnicodeString* id = nullptr;
+            do {
+              id = enumeration->snext(status);
+            } while (U_SUCCESS(status) && --i > 0);
+            if (U_SUCCESS(status)) {
+                return icu::TimeZone::createTimeZone(*id);
+            }
+        }
+    }
+    return icu::TimeZone::getGMT()->clone();
+}
+const char* GetRandomCalendarType(uint16_t rnd) {
+    icu::Locale und("und");
+    UErrorCode status = U_ZERO_ERROR;
+    std::unique_ptr<icu::StringEnumeration> enumeration(
+        icu::Calendar::getKeywordValuesForLocale("calendar", und, false, status));
+    const char* type = "";
+    if (U_SUCCESS(status)) {
+        int32_t count = enumeration->count(status);
+        if (U_SUCCESS(status)) {
+            int32_t i = rnd % count;
+            do {
+              type = enumeration->next(nullptr, status);
+            } while (U_SUCCESS(status) && --i > 0);
+        }
+    }
+    type = uloc_toUnicodeLocaleType("ca", type);
+    return type;
+}
+
+extern "C" int LLVMFuzzerTestOneInput(const uint8_t* data, size_t size) {
+    if (size < 2) return 0;
+    uint16_t rnd16;
+    std::memcpy(&rnd16, data, sizeof(rnd16));
+    data += 2;
+    size -= 2;
+    icu::Locale locale = GetRandomLocale(rnd16);
+    const char* type = GetRandomCalendarType(rnd16);
+    std::unique_ptr<icu::TimeZone> timeZone(CreateRandomTimeZone(rnd16));
```

Comment:
Doesn't this mean you're using the same random number for all three of these values?  Is this really what you want to do?  Maybe it's okay because it's random, but it seems like it'd always yoke together certain combinations of locale, calendar type, and time zone instead of giving you truly random combinations.

Reply:
I share Rich's hunch. I believe that the fuzzing framework uses some kind of guided-random generator that automatically finds interesting combinations of inputs. For that to work well, it seems like a distinct/disjoint input byte sequence should be used for each algorithm input.

Reply:
"the same random number for all three of these values?"
Not exactly. The random number for each are % by the number of combinations for each category therefore are not the same. 

consider the following, if we have X locales, Y calendars and Z timezones.

The locale to be pick is R % X
The calendar to be pick is R % Y
The timezone to be pick is R % Z

As long as X x Y x Z << 2^16 we should hit any combination of locale, calendar, timezone randomly. 
Of course, if X Y and Z share common denomiator then some combination will not be hit. 

We currently have 831 Locales, 18 calendars, and 637 timezones

I can change these lines to 
```
    icu::Locale locale = GetRandomLocale(rnd16 & 0x003ff);  
    const char* type = GetRandomCalendarType( (rnd16 >> 10) & 0x001f ); 
    std::unique_ptr<icu::TimeZone> timeZone(CreateRandomTimeZone((rnd16 >> 15) & 0x003ff );
```
to make sure they pass in muturally exclusive bits from the rnd16.

Reply:
My concern is that those values are still mathematically related and that you might miss combinations as a result, but maybe it's not really a problem in practice.  I like the alternative you suggest better, but I also wonder-- why not have a simple wrapper class around the random number list you can just read random numbers from, like a stream or a generator?  Then you don't have to worry about updating `data` and `size` every time to use a random number, because you're calling a separate class that does that for you.  I think this is basically what Markus is suggesting below, too.

Reply:
ok, I just simplify it and avoid complicate logic.